ARROW-8314: [Python] Add a Table.select method to select a subset of columns

jorisvandenbossche opened the pull request:
This is a pure python implementation. It might be we want that on the C++ side (unless it already exists?), but having it available in Python is already useful IMO.

nealrichardson commented:
Here's a C++ version, albeit from the R bindings: https://github.com/apache/arrow/blob/master/r/src/table.cpp#L128-L151

Since you're doing this in Python as well, maybe this should be moved to C++?

pitrou commented:
@jorisvandenbossche Do you need help on this?

jorisvandenbossche commented:
I added a C++ version (didn't yet update R to use it)

jorisvandenbossche commented on cpp/src/arrow/table.cc:
```diff
@@ -362,6 +362,23 @@ Result<std::shared_ptr<Table>> Table::RenameColumns(
+  }
+
+  auto new_schema =
+      std::make_shared<arrow::Schema>(std::move(fields), schema()->metadata());
```
I am not fully sure here about ownership of the metadata. Or since metadata are immutabe (I think?), passing it as a shared ptr is fine?

pitrou replied:
Yes, metadata is supposed to be immutable, so no point in copying it.

pitrou commented on cpp/src/arrow/table.cc:
```diff
@@ -362,6 +362,23 @@ Result<std::shared_ptr<Table>> Table::RenameColumns(
 
+Result<std::shared_ptr<Table>> Table::SelectColumns(
+    const std::vector<int>& indices) const {
+  int n = indices.size();
```
Need a `static_cast`, or use `int64_t` or `size_t` instead.

pitrou commented:
It would also be nice to add a test on the C++ side, if that's not too time-consuming.

nealrichardson commented:
> I added a C++ version (didn't yet update R to use it)

Are you intending to make that change in this PR too?

jorisvandenbossche commented:
I updated this, and added a small C++ test.

> Are you intending to make that change in this PR too?

Realistically speaking, not at the moment (I certainly want to learn how to set up a R dev environment, but just before the release with other priorities might not be the best time ;-)) 
So would either merge without, unless you can quickly push a change (I suppose it might be an easy change).

nealrichardson commented:
I added https://issues.apache.org/jira/browse/ARROW-9387 for using this in R. It might be trivial but in case it isn't I don't want to block this.

jorisvandenbossche commented:
@pitrou could you check the C++ test?

pitrou commented:
@jorisvandenbossche I'll take a look when I'm done with 1.0.0-critical tasks. Hopefully before the end of the week :-)

wesm commented on cpp/src/arrow/table.cc:
```diff
@@ -362,6 +362,23 @@ Result<std::shared_ptr<Table>> Table::RenameColumns(
+  std::vector<std::shared_ptr<ChunkedArray>> columns(n);
+  std::vector<std::shared_ptr<Field>> fields(n);
+  for (int i = 0; i < n; i++) {
+    int pos = indices[i];
```
If this is returning `Result` anyway we might as well boundscheck the indices, thoughts?

pitrou replied:
+1.

jorisvandenbossche replied:
Added a boundscheck

pitrou commented on cpp/src/arrow/table_test.cc:
```diff
@@ -559,6 +559,17 @@ TEST_F(TestTable, RenameColumns) {
+  MakeExample1(10);
+  auto table = Table::Make(schema_, columns_);
+
+  ASSERT_OK_AND_ASSIGN(auto subset, table->SelectColumns({0, 2}));
```
`ASSERT_OK(subset->ValidateFull())`?

wesm commented:
+1
